File: src/main/java/no/nav/veilarbaktivitet/brukernotifikasjon/avlsutt/AvsluttSender.java
package no.nav.veilarbaktivitet.brukernotifikasjon.avlsutt;

import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import no.nav.brukernotifikasjon.schemas.Done;
import no.nav.brukernotifikasjon.schemas.Nokkel;
import no.nav.common.kafka.producer.KafkaProducerClient;
import no.nav.common.utils.Credentials;
import no.nav.veilarbaktivitet.person.AuthService;
import no.nav.veilarbaktivitet.person.Person;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;

@Service
@RequiredArgsConstructor
class AvsluttSender {
    private final KafkaProducerClient<Nokkel, Done> producer;
    private final AvsluttDao avsluttDao;
    private final AuthService authService;
    private final Credentials serviceUserCredentials;

    @Value("${topic.ut.brukernotifikasjon.done}")
    private String doneToppic;


    @SneakyThrows
    @Transactional
    public void avsluttOppgave(SkalAvluttes skalAvluttes) {
        String aktorId = skalAvluttes.getAktorId();
        String brukernotifikasjonId = skalAvluttes.getBrukernotifikasjonId();

        Person.Fnr fnrForAktorId = authService.getFnrForAktorId(Person.aktorId(aktorId));
        boolean markertAvsluttet = avsluttDao.markerOppgaveSomAvbrutt(brukernotifikasjonId);
        if (markertAvsluttet) {
            Done done = Done
                    .newBuilder()
                    .setFodselsnummer(fnrForAktorId.get())
                    .setGrupperingsId(skalAvluttes.getOppfolgingsPeriode().toString())
                    .setTidspunkt(Instant.now().toEpochMilli())
                    .build();

            Nokkel nokkel = new Nokkel(serviceUserCredentials.username, brukernotifikasjonId);
            final ProducerRecord<Nokkel, Done> kafkaMelding = new ProducerRecord<>(doneToppic, nokkel, done);

            producer.send(kafkaMelding).get();
        }
    }

    public List<SkalAvluttes> getOppgaverSomSkalAvbrytes(int maxAntall) {
        return avsluttDao.getOppgaverSomSkalAvsluttes(maxAntall);
    }

    public int avsluttIkkeSendteOppgaver() {
        return avsluttDao.avsluttIkkeSendteOppgaver();
    }
}

File: src/main/kotlin/no/nav/veilarbaktivitet/brukernotifikasjon/opprettVarsel/AvsluttSender.kt
package no.nav.veilarbaktivitet.brukernotifikasjon.opprettVarsel

import io.micrometer.core.annotation.Timed
import lombok.RequiredArgsConstructor
import lombok.SneakyThrows
import lombok.extern.slf4j.Slf4j
import no.nav.tms.varsel.action.Produsent
import no.nav.tms.varsel.builder.VarselActionBuilder
import no.nav.veilarbaktivitet.brukernotifikasjon.avslutt.AvsluttDao
import no.nav.veilarbaktivitet.brukernotifikasjon.avslutt.SkalAvluttes
import no.nav.veilarbaktivitet.config.kafka.kafkatemplates.KafkaStringTemplate
import org.apache.kafka.clients.producer.ProducerRecord
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@RequiredArgsConstructor
@Slf4j
internal open class AvsluttSender(
    private val producer: KafkaStringTemplate,
    private val avsluttDao: AvsluttDao,
    @Value("\${topic.ut.brukernotifikasjon.brukervarsel}")
    private val brukervarselTopic: String,
    @Value("\${app.env.appname}")
    private val appname: String,
    @Value("\${app.env.namespace}")
    private val namespace: String,
    @Value("\${app.env.cluster}")
    private val cluster: String
) {

    private val log = LoggerFactory.getLogger(AvsluttSender::class.java)

    @SneakyThrows
    @Transactional
    @Timed(value = "brukernotifikasjon_avslutt_oppgave_sendt")
    open fun avsluttOppgave(skalAvluttes: SkalAvluttes) {
        val brukernotifikasjonId = skalAvluttes.brukernotifikasjonId

        val markertAvsluttet = avsluttDao.markerOppgaveSomAvsluttet(brukernotifikasjonId)
        if (markertAvsluttet) {
            val melding = VarselActionBuilder.inaktiver {
                this.varselId = brukernotifikasjonId
                this.produsent = Produsent(
                    cluster = cluster,
                    namespace = namespace,
                    appnavn = appname,
                )
            }
            val kafkaMelding = ProducerRecord(brukervarselTopic, skalAvluttes.brukernotifikasjonId, melding)
            log.info(
                "Sender brukernotifikasjon 'done' for grupperingsid: {}",
                skalAvluttes.getOppfolgingsperiode().toString()
            )
            producer.send(kafkaMelding).get()
        }
    }

    open fun getOppgaverSomSkalAvbrytes(maxAntall: Int): List<SkalAvluttes> {
        return avsluttDao.getOppgaverSomSkalAvsluttes(maxAntall)
    }

    open fun avsluttIkkeSendteOppgaver(): Int {
        return avsluttDao.avsluttIkkeSendteOppgaver()
    }

    open fun markerAvslutteterAktiviteterSomSkalAvsluttes(): Int {
        return avsluttDao.markerAvslutteterAktiviteterSomSkalAvsluttes()
    }
}
